integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/multitenancy/DatabaseMultitenancyIT.java

```diff
@@ -0,0 +1,180 @@
+import org.hibernate.search.util.impl.integrationtest.common.rule.BackendMock;
+import org.hibernate.search.util.impl.integrationtest.common.rule.StubSearchWorkBehavior;
+import org.hibernate.search.util.impl.integrationtest.mapper.orm.OrmSetupHelper;
+import org.hibernate.search.util.impl.test.annotation.PortedFromSearch5;
+import org.hibernate.search.util.impl.test.annotation.TestForIssue;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+@TestForIssue(jiraKey = "HSEARCH-4034")
+@PortedFromSearch5(original = "org.hibernate.search.test.batchindexing.DatabaseMultitenancyTest")
+public class DatabaseMultitenancyIT {
+
+	@Rule
+	public BackendMock backendMock = new BackendMock();
+
+	@Rule
+	public OrmSetupHelper ormSetupHelper = OrmSetupHelper.withBackendMock( backendMock );
+
+	/**
+	 * Metamec tenant identifier
+	 */
+	private static final String METAMEC_TID = "metamec";
+
+	/**
+	 * Geochron tenant identifier
+	 */
+	private static final String GEOCHRON_TID = "geochron";
+
+	private static Clock[] METAMEC_MODELS = {
+			new Clock( 1, "Metamec - Model A850" ),
+			new Clock( 2, "Metamec - Model 4562" ),
+			new Clock( 5, "Metamec - Model 792" )
+	};
+
+	private static Clock[] GEOCHRON_MODELS = {
+			new Clock( 1, "Geochron - Model The Original Kilburg" ),
+			new Clock( 2, "Geochron - Model The Boardroom" ),
+			new Clock( 9, "Geochron - Model Designer Series" )
+	};
+
+	private SessionFactory sessionFactory;
+
+	@Before
+	public void setup() {
+		backendMock.expectSchema( Clock.INDEX, b -> b.field( "brand", String.class ) );
+		sessionFactory = ormSetupHelper
+				.start()
+				.tenants( METAMEC_TID, GEOCHRON_TID )
+				.setup( Clock.class );
+		backendMock.verifyExpectationsMet();
+
+		// init data:
+		persist( METAMEC_TID, METAMEC_MODELS );
+		persist( GEOCHRON_TID, GEOCHRON_MODELS );
+	}
+
+	@Test
+	public void shouldOnlyFindMetamecModels() {
+		List<Clock> list = searchAll( METAMEC_TID, METAMEC_MODELS );
+		assertThat( list ).containsExactlyInAnyOrder( METAMEC_MODELS );
+	}
+
+	@Test
+	public void shouldOnlyFindGeochronModels() {
+		List<Clock> list = searchAll( GEOCHRON_TID, GEOCHRON_MODELS );
+		assertThat( list ).containsExactlyInAnyOrder( GEOCHRON_MODELS );
+	}
+
+	@Test
+	public void shouldMatchOnlyElementsFromOneTenant() {
+		List<Clock> list = searchModel( "model", GEOCHRON_TID, GEOCHRON_MODELS );
+		assertThat( list ).containsExactlyInAnyOrder( GEOCHRON_MODELS );
+	}
+
+	@Test
+	public void searchReferences() {
+		List<EntityReference> entityReferences = searchReferences( GEOCHRON_TID, GEOCHRON_MODELS );
+		assertThat( entityReferences ).hasSize( GEOCHRON_MODELS.length );
+	}
+
+	private void persist(String tenantId, Clock[] models) {
+		BackendMock.DocumentWorkCallListContext expectWorks = backendMock.expectWorks( Clock.INDEX );
+		for ( Clock clock : models ) {
+			expectWorks.add( clock.getId() + "", tenantId, b -> b.field( "brand", clock.getBrand() ) );
```

Comment:
Wouldn't it make more sense to pass the tenantId to `expectWorks()`, just above the `for` loop? That's what we do for `expectIndexScaleWorks`, at least.

Reply:
Yes it is better to set the tenant on `DocumentWorkCallListContext` once and for all